apps/src/applab/RecordListener.js

```diff
@@ -0,0 +1,230 @@
+   * @private {Object.<string, TableHandler>}
+   */
+  this.tableHandlers_ = {};
+};
+
+/**
+ * Number of ms to wait before polling each table for more data after finishing
+ * processing the last batch of data for that table.
+ * @type {number}
+ */
+var RECORD_INTERVAL = 1000; // 1 second
+
+var EventType = {
+  CREATE: 'create',
+  UPDATE: 'update',
+  DELETE: 'delete'
+};
+RecordListener.EventType = EventType;
+
+/**
+ * Adds a listener which calls the callback when changes are made to the table.
+ * The callback is called once per changed record.
+ * @param {string} tableName Name of the table to listen to.
+ * @param {function(Object, EventType)} callback Callback to call with the record
+ * and the type of event which happened to it. For create and update events,
+ * the entire new record is included. For delete events, only the id is included.
+ * @returns {boolean} true if adding the listener succeeded, or false if
+ * a listener already existed for the specified table.
+ */
+RecordListener.prototype.addListener = function (tableName, callback) {
+  if (this.tableHandlers_[tableName]) {
+    return false;
+  }
+
+  this.tableHandlers_[tableName] = new TableHandler(tableName, callback);
+
+  return true;
+};
+
+/**
+ * Clears the timeouts and resets internal state.
+ */
+RecordListener.prototype.reset = function () {
+  for (var tableName in this.tableHandlers_) {
+    this.tableHandlers_[tableName].reset();
+  }
+  this.tableHandlers_ = {};
+};
+
+//////////////////////////////////////////////////
+// TableHandler
+//////////////////////////////////////////////////
+
+/**
+ * This class encapsulates all the logic and state required for listening to
+ * changes to a single table.
+ * @param {string} tableName
+ * @param {function} callback
+ * @constructor
+ */
+var TableHandler = function (tableName, callback) {
+  /**
+   * @private {string}
+   */
+  this.tableName_ = tableName;
+
+  /**
+   *  @private {function}
+   */
+  this.callback_ = callback;
+
+  /**
+   * @typedef {Object.<number, string>} IdToJsonMap Map from record id
+   * to a JSON-stringified copy of the corresponding record.
+   */
+
+  /**
+   * IdToJsonMap representing the contents of the table.
+   * We use this to efficiently test whether a record's contents has changed.
+   * @private {IdToJsonMap}
+   */
+  this.idToJsonMap_ = {};
+
+  /**
+   * Timeout id of the pending call to window.setTimeout.
+   * @private {number}
+   */
+  this.timeoutId_ = null;
+
+  this.scheduleNextFetch_();
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.scheduleNextFetch_ = function () {
+  this.timeoutId_ = setTimeout(function recordListener() {
+    this.fetchRecords_();
+  }.bind(this), RECORD_INTERVAL);
+};
+
+/**
+ * @private
+ */
+TableHandler.prototype.fetchRecords_ = function () {
+  var req = new XMLHttpRequest();
+  req.onreadystatechange = this.handleFetchRecords_.bind(this, req);
+  var url = '/v3/shared-tables/' + Applab.channelId + '/' + this.tableName_;
```

Comment:
Are there other places where we back URLs like this?  We should factor out a helper for constructing the URL to avoid duplicating knowledge of how it is formatted.